### PR TITLE
[Experimental] DataProfiler Expectations

### PR DESCRIPTION
**Pull request by taylorfturner**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:

Add two new expectations inspired by Capital One's [DataProfiler](https://github.com/capitalone/DataProfiler) package
- Expect all column value model confidencs to be `>=` `x` threshold: `expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py`
     - This expectation builds upon the cusotm map expectation of Great Expectations. This function asks the following yes/no question: are each confidence values of the user-specified column great than or equal to `x` threshold. 
     - Example: ```df.expect_column_values_to_probabilistically_match_data_label(
        column,
        data_label=<>,
        threshold=float(0<=1))```

- Expect the column value confidences for a give data label to be `>-` `x` threshold:`expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py`
     - This expectation builds upon the custom column map expectation of Great Expectations. This function asks the yes/no question of each row in the user-specified column; namely, does the confidence threshold provided by the DataProfiler model exceed the user-specified threshold. 
     - Example: ```df.expect_column_values_to_probabilistically_match_data_label(
        column,
        threshold=float(0<=1))```

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by taylorfturner** [comment]
@cla-bot check

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Implement the yes/no question for the expectation
+        """
+        labeler = dp.DataLabeler(labeler_type='structured')
+        try:
+            results = labeler.predict(column, predict_options={"show_confidences": True})
+        except:
+            results = None
+        label_map_vec_func = np.vectorize(lambda x: labeler.label_mapping.get(x, None))
+        results['pred'] = label_map_vec_func(results['pred'])
```
Comment (JGSweets) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py]: Should not need this, can just tell the postprocessor to use the index output format

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        Implement the yes/no question for the expectation
+        """
+        labeler = dp.DataLabeler(labeler_type='structured')
+        try:
+            results = labeler.predict(column, predict_options={"show_confidences": True})
+        except:
+            results = None
+        
+        label_map_vec_func = np.vectorize(lambda x: labeler.label_mapping.get(x, None))
+        results['pred'] = label_map_vec_func(results['pred'])
```
Comment (JGSweets) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]: as above, utilize the postprocessor parameters to do this auto.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    SqlAlchemyExecutionEngine,
+)
+from great_expectations.expectations.expectation import ColumnMapExpectation
+from great_expectations.expectations.metrics import (
+    ColumnMapMetricProvider,
+    column_condition_partial,
+)
+
+
+class ColumnValuesConfidenceForDataLabelToBeGreaterThanOrLessThanThreshold(ColumnMapMetricProvider):
```
Comment (JGSweets) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]: `GreaterThanOrLessThanThreshold` -> `GreaterThanOrEqualThreshold`?

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        
+        label_map_vec_func = np.vectorize(lambda x: labeler.label_mapping.get(x, None))
+        results['pred'] = label_map_vec_func(results['pred'])
+        test_int_data_label = label_map_vec_func(data_label)
+
+        # list comprehension to only select the confidence values for a given data label 
+        #   within the column specified
+        results_array = np.array([confidences[results['pred'][iter_value]]\
+            if results['pred'][iter_value] == test_int_data_label else .00\
+                for iter_value, confidences in enumerate(results['conf'])])
```
Comment (JGSweets) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]:
```python
data_label_ind = labeler.label_mapping[data_label]  # will need error checking around it
data_label_conf = results['conf'][:, data_label_ind]
return data_label_conf >= threshold
````

Comment (JGSweets) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]: can also do:
```python
data_label_ind = labeler.label_mapping.get(data_label, None)
if data_label_ind is None:
    raise ....
```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    @column_condition_partial(engine=PandasExecutionEngine)
+    def _pandas(cls, column, threshold, data_label, **kwargs):
+        """
+        Implement the yes/no question for the expectation
+        """
+        labeler = dp.DataLabeler(labeler_type='structured')
+        try:
+            results = labeler.predict(column, predict_options={"show_confidences": True})
+        except:
+            results = None
```
Comment (JGSweets) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]: How should this be handled?

Comment (taylorfturner) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py]: I actually lean towards taking this out entirely. I did add a `raise` on the `data_label` parameter so its more explicit what a user can submit. But this error should be handled by DP or Great Expectations when it runs.

**Review by talagluck — COMMENTED** [review]
This is awesome! I'll dive in and test this more thoroughly bit later, but at first glance I see a couple of minor things:

- It looks like our pipeline tests are failing due to linting. Would you please run a linter [as described here](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)?

- I love that you have the input validation for the `column_values_confidence` expectation. Another way you could handle this is with a `validate_configuration` method within the Expectation, [similar to this.](https://github.com/great-expectations/great_expectations/blob/7b23728d8f282d997eba41767bff5d5491a88779/great_expectations/expectations/core/expect_column_values_to_be_in_set.py#L356)

Thanks!

**Comment by taylorfturner** [comment]
> This is awesome! I'll dive in and test this more thoroughly bit later, but at first glance I see a couple of minor things:
> 
> * It looks like our pipeline tests are failing due to linting. Would you please run a linter [as described here](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)?
> * I love that you have the input validation for the `column_values_confidence` expectation. Another way you could handle this is with a `validate_configuration` method within the Expectation, [similar to this.](https://github.com/great-expectations/great_expectations/blob/7b23728d8f282d997eba41767bff5d5491a88779/great_expectations/expectations/core/expect_column_values_to_be_in_set.py#L356)
> 
> Thanks!

I'll take a look at these right now.

**Comment by taylorfturner** [comment]
Pending additional comments from the Great Expectations team...

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        )
+        return (
+            np.choose(results["pred"].astype(int, copy=False), results["conf"].T)
+            >= threshold
+        )
+
+
+class ExpectColumnValuesToBeGreaterThanOrEqualToThreshold(ColumnMapExpectation):
+    """
+    This function builds upon the custom column map expectations of Great Expectations. This function asks the question a yes/no question of each row in the user-specified column; namely, does the confidence threshold provided by the DataProfiler model exceed the user-specified threshold.
```
Comment (talagluck) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py]:
```suggestion
    This function builds upon the custom column map expectations of Great Expectations. This function asks a yes/no question of each row in the user-specified column; namely, does the confidence threshold provided by the DataProfiler model exceed the user-specified threshold.
```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    success_keys = (
+        "threshold",
+        "mostly",
+    )
+
+    # This dictionary contains default values for any parameters that should have default values
+    default_kwarg_values = {}
+
+    # This object contains metadata for display in the public Gallery
+    library_metadata = {
```
Comment (talagluck) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py:155]: @taylorfturner - could you please add a `requirements` key to the the `library_metadata` object with a list of any additional dependencies? It looks like in addition to `dataprofiler`, `tensorflow` and `scikit-learn` are requirements. [You can see an example here.](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_us_zipcode_within_radius_of_given_zipcode.py#L245). You'll also see that it is included in the docstring on L38.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        results = labeler.predict(
+            column, predict_options={"show_confidences": True}
+        )
+        return (
+            np.choose(results["pred"].astype(int, copy=False), results["conf"].T)
+            >= threshold
+        )
+
+
+class ExpectColumnValuesToBeGreaterThanOrEqualToThreshold(ColumnMapExpectation):
```
Comment (talagluck) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py:55]: To put it in simpler terms, is this Expectation about how effectively DataProfiler can Label a column, irrespective of the specifics of the Label or column contents?

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        "threshold",
+        "data_label",
+        "mostly",
+    )
+
+    # This dictionary contains default values for any parameters that should have default values
+    default_kwarg_values = {}
+
+    # This object contains metadata for display in the public Gallery
+    library_metadata = {
```
Comment (talagluck) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py:174]: As above, would you please add a `requirements` key here?

**Review by talagluck — COMMENTED** [review]
This looks great overall! Just a few additional notes within.

Additionally, for `expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py`, would you please add an "or" in between "than" and "equal". It might also be fair to remove the "or equal to" from the Expectation name, based on some other Expectations that we have (e.g. `expect_column_kl_divergence_to_be_less_than`, `expect_column_chisquare_test_p_value_to_be_greater_than`). In that case, you may want an addition flag for the "or equal to" condition.

This would add some complexity, so leaving that decision up to you.

We are also currently in the process of finalizing our workflow for Expectation packages, and would love for you to test that out. The idea would basically be that instead of contributing these to the general list of `contrib` expectations, we would have a `Capital One DataProfiler` package, which would include any and all Expectations in this vein (and could be added to down the road).

I'll be out next week, but @austiezr will be in touch to help finish up the PR and share more information around the packaging process.

**Comment by taylorfturner** [comment]
@talagluck and @austiezr is this a known issue on the repo with regard to `black .`

```error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/create_expectations.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/create_expectations.ipynb' cannot be parsed as valid Jupyter notebook.
error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/create_expectations.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/create_expectations.ipynb' cannot be parsed as valid Jupyter notebook.
error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.
error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.
error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/create_expectations.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/create_expectations.ipynb' cannot be parsed as valid Jupyter notebook.
error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.```

**Comment by austiezr** [comment]
> @talagluck and @austiezr is this a known issue on the repo with regard to `black .`
> 
> ```
> error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/create_expectations.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/create_expectations.ipynb' cannot be parsed as valid Jupyter notebook.
> error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/spark/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.
> error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/pandas/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.
> error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/create_expectations.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/create_expectations.ipynb' cannot be parsed as valid Jupyter notebook.
> error: cannot format tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/validation_playground.ipynb: File 'tests/data_context/fixtures/post_init_project_v0.8.0_A/great_expectations/notebooks/sql/validation_playground.ipynb' cannot be parsed as valid Jupyter notebook.```
> ```

Hey @taylorfturner ! Thanks for surfacing this; we'll investigate. Go ahead and ignore!

**Comment by taylorfturner** [comment]
Very good -- will do, @austiezr

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py (hunk 8 above)
Comment (taylorfturner) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py:55]: correct

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py (hunk 9 above)
Comment (taylorfturner) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_confidence_to_be_greater_than_or_equal_to_threshold_for_data_label.py:174]: got it

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py (hunk 7 above)
Comment (taylorfturner) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_probabilistically_greater_than_equal_to_threshold.py:155]: got it

**Comment by taylorfturner** [comment]
**I have closed this PR and opening a new PR [here](https://github.com/great-expectations/great_expectations/pull/4174) to make this is a bit clean**

@austiezr there appears to be an issue with the remote linting. It's failing on [this](https://dev.azure.com/great-expectations/great_expectations/_build/results?buildId=30520&view=logs&j=eca830a7-e027-5c16-85fb-144337d17bc8&t=0fd6b91e-dc3f-5154-ed9f-0aea14fe2f7b) 

Also in response to @talagluck's comment, I could use some additional thoughts and ideas around this. Would this be completely outside of the `great_expectations` repository or just a new sub-folder within the package? 
> We are also currently in the process of finalizing our workflow for Expectation packages, and would love for you to test that out. The idea would basically be that instead of contributing these to the general list of contrib expectations, we would have a Capital One DataProfiler package, which would include any and all Expectations in this vein (and could be added to down the road).

**Comment by JGSweets** [comment]
> We are also currently in the process of finalizing our workflow for Expectation packages, and would love for you to test that out. The idea would basically be that instead of contributing these to the general list of `contrib` expectations, we would have a `Capital One DataProfiler` package, which would include any and all Expectations in this vein (and could be added to down the road).

Also curious how this would function. In the DataProfiler, we solved this by utilizing pip's capability to alter optional dependency with `[]` notation, e.g. `pip install dataprofiler[ml]`.

I think our reports module did it best, where we utilize a decorator on specific functions to alert the user of extra requirements they could install to utilize those functions:
https://github.com/capitalone/DataProfiler/blob/main/dataprofiler/reports/graphs.py#L20

However, the imports still need protecting to ensure no error on load:
https://github.com/capitalone/DataProfiler/blob/main/dataprofiler/reports/graphs.py#L7

Technically, one could move the imports to inside the function as well. However, imports become less clear IMO.
